### PR TITLE
Create `ImageFilter.dilate`/`ImageFilter.erode`

### PR DESCRIPTION
**Pull request by fzyzcjy**
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*
Please see https://github.com/flutter/flutter/issues/100830

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/issues/100830

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by fzyzcjy** [comment]
May I know where to add tests for Dart side?

**Comment by fzyzcjy** [comment]
The failing `linux unopt` does not seem to be related to me?

```
   info - canvas_test.dart:269:5 - Prefer const over final for declarations. - prefer_const_declarations
   info - canvas_test.dart:270:5 - Prefer const over final for declarations. - prefer_const_declarations
   info - fragment_shader_test.dart:110:7 - Specify type annotations. - always_specify_types
   info - fragment_shader_test.dart:111:7 - Specify type annotations. - always_specify_types
   info - fragment_shader_test.dart:117:7 - Specify type annotations. - always_specify_types
   info - fragment_shader_test.dart:118:7 - Specify type annotations. - always_specify_types
   info - fragment_shader_test.dart:131:5 - Specify type annotations. - always_specify_types
   info - fragment_shader_test.dart:132:5 - Specify type annotations. - always_specify_types
   info - fragment_shader_test.dart:254:13 - The declaration '_blueColor' isn't referenced. Try removing the declaration of '_blueColor'. - unused_element
   info - fragment_shader_test.dart:267:3 - Prefer const over final for declarations. - prefer_const_declarations
   info - fragment_shader_test.dart:268:3 - Prefer const over final for declarations. - prefer_const_declarations
   info - fragment_shader_test.dart:294:19 - The declaration '_singleUniform' isn't referenced. Try removing the declaration of '_singleUniform'. - unused_element
   info - image_shader_test.dart:5:8 - Unused import: 'dart:async'. Try removing the import directive. - unused_import
   info - image_shader_test.dart:17:23 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:73:22 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:78:43 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:86:41 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:94:44 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:102:42 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:122:22 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:127:43 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:135:41 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:143:44 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:151:42 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:171:22 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:176:44 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:185:52 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:193:49 - Prefer const with constant constructors. - prefer_const_constructors
   info - paragraph_test.dart:201:50 - Prefer const with constant constructors. - prefer_const_constructors
   info - serial_gc_test.dart:11:10 - Prefer final for variable declarations if they are not reassigned. - prefer_final_locals
   info - serial_gc_test.dart:13:7 - Specify type annotations. - always_specify_types
   info - serial_gc_test.dart:13:11 - Prefer final for variable declarations if they are not reassigned. - prefer_final_locals
   info - serial_gc_test.dart:13:11 - The value of the local variable 'a' isn't used. Try removing the variable or using it. - unused_local_variable
   info - text_test.dart:255:26 - Prefer const with constant constructors. - prefer_const_constructors
   info - text_test.dart:265:39 - Prefer const with constant constructors. - prefer_const_constructors
   info - text_test.dart:269:26 - Prefer const with constant constructors. - prefer_const_constructors

36 issues found.
```

I did not change them...

File: lib/ui/painting.dart
Comment (dnfield) [inline lib/ui/painting.dart:3352]: remove comment

File: lib/ui/painting.dart
Comment (dnfield) [inline lib/ui/painting.dart:3382]: remove

File: lib/web_ui/lib/painting.dart
Comment (dnfield) [inline lib/web_ui/lib/painting.dart:419]: Please file or link to existing github issues for this.

The CanvasKit implementations shouldn't be too hard. The HTML backend might be harder.

**Review by dnfield — COMMENTED** [review]
Just some nits, would defer to @flar for approval on DL code and tests but it all seems pretty fine to me.

@bdero @chinmaygarde FYI since this increases surface API that Impeller will end up needing to support if we land it.

File: display_list/display_list_image_filter.h
Comment (flar) [inline display_list/display_list_image_filter.h:271]: Again, no `* 3` here.

File: lib/ui/painting.dart
Comment (flar) [reply lib/ui/painting.dart:3352]: Or correct it to the right name for consistency with the other _Filter objects. Personally I don't see the value in the comment, so not propagating that consistency makes sense as well.

File: lib/ui/painting.dart
Comment (flar) [reply lib/ui/painting.dart:3382]: Same comment for me as above. There is value in consistency with the other _Filter objects, but there is also value in not having obvious one-liner comments.

File: lib/ui/painting.dart
Comment (flar) [inline lib/ui/painting.dart:3228]: Is there a better description than "dilates each input pixel's ...". You mentioned a similarity with blur - does it blur them over that radius? Does it add them? Or does it propagate/duplicate them? And how does that action differ for erode? Is erode a subtraction?

File: lib/ui/painting.dart
Comment (flar) [inline lib/ui/painting.dart:3235]: Same comment as above - can we be more descriptive about the operation for our other developers who aren't already sure that they need this operation.

File: lib/web_ui/lib/painting.dart
Comment (flar) [reply lib/web_ui/lib/painting.dart:419]: Yes, please look into implementations here if at all possible. I know it's a fairly big additional body of code to sift through, but we should make it a priority not to have the DOM/canvaskit implementations fall behind as we add new features.

CC'ing @yjbanov

File: testing/dart/image_filter_test.dart
Comment (flar) [inline testing/dart/image_filter_test.dart:191]: This makes it look like the operation is just a fill operation. Is there a way we can make this test more robust?

Comment (flar) [reply testing/dart/image_filter_test.dart:191]: I suppose this is showing the effect as it copied (dilated) the center green pixel out to the sides and corners (this seems odd that the parameters are called radii then as that would imply a circular dilation, but this seems to be doing a box dilation?)

Comment (flar) [reply testing/dart/image_filter_test.dart:191]: I guess the term radius comes from SVG which uses that name even though the effect is a box filter.

**Review by flar — CHANGES_REQUESTED** [review]
It looks good with a few minor fixes.

File: display_list/display_list_image_filter.h
Comment (fzyzcjy) [reply display_list/display_list_image_filter.h:225]: I will fix that for blur as well.

File: lib/ui/painting.dart
Comment (fzyzcjy) [reply lib/ui/painting.dart:3228]: Well I copied the comments from skia's doc since I think that is the most brief and official.

File: testing/dart/image_filter_test.dart
Comment (fzyzcjy) [reply testing/dart/image_filter_test.dart:191]: > I suppose this is showing the effect as it copied (dilated) the center green pixel out to the sides and corners (this seems odd that the parameters are called radii then as that would imply a circular dilation, but this seems to be doing a box dilation?)

Yes I guess it is box dilation. The reason may be, box is much, much faster than circular dilation (as I have learnt in computer vision courses).

File: lib/web_ui/lib/painting.dart
Comment (fzyzcjy) [reply lib/web_ui/lib/painting.dart:419]: HTML backend seems nontrivial. I see some people use svg to do so, but I do not think it is valid for this context. Not find any solutions with CSS (unlike blur - which has CSS implementation).

Comment (fzyzcjy) [reply lib/web_ui/lib/painting.dart:419]: In addition, I guess the `compose` is also not implemented:

  // TODO(ferhat): add implementation and remove the "ignore".
  // ignore: avoid_unused_constructor_parameters
  ImageFilter.compose({required ImageFilter outer, required ImageFilter inner}) {
    throw UnimplementedError(
        'ImageFilter.compose not implemented for web platform.');
  
}

Comment (fzyzcjy) [reply lib/web_ui/lib/painting.dart]: Github link added.

File: display_list/display_list_canvas_unittests.cc
Comment (fzyzcjy) [reply display_list/display_list_canvas_unittests.cc]: Seems yes. At least for erode, you need a big stroke width. Otherwise, when you erode, you will get nothing left, and tests will complain that no pixels are touched.

**Comment by fzyzcjy** [comment]
The error seems to be:

```
02:50 [32m+259[0m[33m ~6[0m[31m -1[0m: test/html/paragraph/bidi_golden_test.dart: bidi with selection (DOM) [1m[31m[E][0m[0m                                                                                                             
  TimeoutException after 0:00:30.000000: Test timed out after 30 seconds. See https://pub.dev/packages/test#timeouts
  /Users/chrome-bot/.pub-cache/hosted/pub.dartlang.org/test_api-0.4.1/lib/src/backend/invoker.dart 333:9   call
  org-dartlang-sdk:///lib/async/zone.dart 1418:39                                                          _rootRun
  org-dartlang-sdk:///lib/async/zone.dart 1327:34                                                          run
  /Users/chrome-bot/.pub-cache/hosted/pub.dartlang.org/test_api-0.4.1/lib/src/backend/invoker.dart 331:5   _handleError
  /Users/chrome-bot/.pub-cache/hosted/pub.dartlang.org/test_api-0.4.1/lib/src/backend/invoker.dart 326:8   _handleError
  /Users/chrome-bot/.pub-cache/hosted/pub.dartlang.org/test_api-0.4.1/lib/src/backend/invoker.dart 287:9   call
  org-dartlang-sdk:///lib/async/zone.dart 1426:12                                                          _rootRun
  org-dartlang-sdk:///lib/async/zone.dart 1327:34                                                          run
  /Users/chrome-bot/.pub-cache/hosted/pub.dartlang.org/test_api-0.4.1/lib/src/backend/invoker.dart 286:38  call
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/async_patch.dart 131:9                                  call
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart 1818:14                                  invokeClosure
  org-dartlang-sdk:///lib/_internal/js_runtime/lib/js_helper.dart 1852:1                                   <fn>
  
Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'pub run test --chain-stack-traces'.
```

Thus maybe unrelated to my PR?

**Comment by fzyzcjy** [comment]
By the way, I see `25. Upload artifacts android-arm64-profile` in the CI. So I wonder where is the uploaded artifacts? I want to have a try on them to ensure they are really correct when used in my app.

**Comment by fzyzcjy** [comment]
Hi, is there any updates?

**Comment by dnfield** [comment]
There were some CI issues today or yesterday - not sure if they're resolved yet but might be worth just merging up to head and pushing a new commit to kick CI.

**Comment by fzyzcjy** [comment]
@flar @dnfield It is green now!

File: lib/ui/painting.dart
Comment (flar) [reply lib/ui/painting.dart:3228]: For the record, Flutter has a different audience than Skia. In this case our overall ImageFilter documentation is pretty sparse overall anyway, so this should probably be addressed  in a sweep.

Flutter developers are probably more aware of CSS or SVG specs and concepts than "a famous computer vision library".

File: display_list/display_list_canvas_unittests.cc
Comment (flar) [inline display_list/display_list_canvas_unittests.cc:1171]: Thank you for updating these tests, BTW, many of the added capabilities here don't really add tests to this file. It's not a hard requirement, but one thing it does is to figure out if we're computing accurate bounds.

To that end, the tolerance object is primarily meant to indicate "the bounds you compute may be conservative and it is too hard to come up with a test case that demonstrates the pessimal bounds case that drives the conservative computations.

In this case, was 21 padding really needed? The bounds would be enlarged by only 1 pixel on each side and the bounds computation on this type of filter should be really accurate. I'm surprised any bounds padding is needed at all. Was this perhaps from the time when DlDilateFilter was computing bounds with an extra `* 3`?

File: display_list/display_list_canvas_unittests.cc
Comment (flar) [inline display_list/display_list_canvas_unittests.cc:1140]: See comment below regarding bounds padding.

File: display_list/display_list_canvas_unittests.cc
Comment (flar) [inline display_list/display_list_canvas_unittests.cc:1174]: You call this "Erode 3", but the parameters are 1.0 above.

File: display_list/display_list_image_filter_unittests.cc
Comment (flar) [inline display_list/display_list_image_filter_unittests.cc:61]: There are too calls verifying asDilate and non testing asBlur. I think some cutting/pasting went awry here.

File: display_list/display_list_image_filter_unittests.cc
Comment (flar) [inline display_list/display_list_image_filter_unittests.cc:77]: Three calls to asErode. Cut/paste error?

**Review by flar — APPROVED** [review]
Looks good to go pending a couple of minor changes:

- Fix what looks like cut/paste errors in the dl_unittests.cc
- Try to pare down the reliance on bounds padding in the canvas_unittests.cc (for Dilate/Erode the bounds should be nearly perfectly accurate so a padding tolerance of 21 pixels seems a bit overkill, hopefully the dilate/erode tests would not need bounds tolerance at all?)

**Comment by fzyzcjy** [comment]
Everything seems green now, what should I do next?

image

**Comment by flar** [comment]
> Everything seems green now, what should I do next?
> 
> image

In general, we are moving to a state where all PRs have to use the "Waiting for tree to go green" label to get submitted. I seem to recall a requirement that developers need 2 approving reviews from authorized reviewers to be eligible to be pushed, but I don't see the checks asking for that. I'll add the label and see if it goes in.

**Comment by flar** [comment]
If the bot comes along and removes the label, then tag @dnfield to complete a review and then the label should work.

Note that for now the tree is broken (the "luci-engine" failure in the list) so the label is currently in a wait state on that condition.

**Comment by fzyzcjy** [comment]
@flar I see.

@dnfield Hi could you please approve this PR such that it can be merged?

**Review by dnfield — APPROVED** [review]

**Comment by dnfield** [comment]
Once the tree is open this should get landed by the bot.

**Comment by fzyzcjy** [comment]
@dnfield Thank you

**Pull request merged**
